adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,928 @@
+
+_Effects (2)_: Equivalent to [code]#h.fill(dest, src)#.
+
+_Effects (3)_: Equivalent to [code]#q.submit([&](handler& h) { fill(h, ptr,
+pattern, count); })#.
+
+_Effects (4)_: Equivalent to [code]#q.submit([&](handler& h) { fill(h, dest,
+src); })#.
+
+'''
+
+.[apititle]#update_host#
+[source,role=synopsis,id=api:update_host]
+----
+namespace sycl::khr {
+
+template <typename T, int Dims, access_mode Mode>
+void update_host(sycl::handler& h, accessor<T, Dims, Mode, target::device> acc); (1)
+
+template <typename T, int Dims, access_mode Mode>
+void update_host(sycl::queue q, accessor<T, Dims, Mode, target::device> acc);    (2)
+
+}
+----
+_Constraints_: [code]#T# must be <<device-copyable>>.
+
+_Effects (1)_: Equivalent to [code]#h.update_host(acc)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { update_host(h,
+acc); })#.
+
+'''
+
+.[apititle]#prefetch#
+[source,role=synopsis,id=api:prefetch]
+----
+namespace sycl::khr {
+
+void prefetch(sycl::handler& h, void* ptr, size_t numBytes); (1)
+
+void prefetch(sycl::queue q, void* ptr, size_t numBytes);    (2)
+
+}
+----
+_Effects (1)_: Equivalent to [code]#h.prefetch(ptr, numBytes)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { prefetch(h, ptr,
+numBytes); })#.
+
+'''
+
+.[apititle]#mem_advise#
+[source,role=synopsis,id=api:mem_advise]
+----
+namespace sycl::khr {
+
+void mem_advise(sycl::handler& h, void* ptr, size_t numBytes, int advice); (1)
+
+void mem_advise(sycl::queue q, void* ptr, size_t numBytes, int advice);    (2)
+
+}
+----
+_Effects (1)_: Equivalent to [code]#h.mem_advise(ptr, numBytes, advice)#.
+
+_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { mem_advise(h,
+ptr, numBytes, advice); })#.
+
+'''
+
+=== Command and event barriers
+
+.[apititle]#command_barrier#
+[source,role=synopsis,id=api:command_barrier]
+----
+namespace sycl::khr {
+
+void command_barrier(sycl::handler& h); (1)
+
+void command_barrier(sycl::queue q);    (2)
+
+}
+----
+_Effects_: Enqueues a command barrier.
+Any commands submitted after this barrier cannot begin execution until all
+previously submitted commands (and any commands associated with dependendent
+events) have completed.
+
+'''
+
+.[apititle]#event_barrier#
+[source,role=synopsis,id=api:event_barrier]
+----
+namespace sycl::khr {
+
+void event_barrier(sycl::handler& h, const std::vector<sycl::event>& events); (1)
+
+void event_barrier(sycl::queue q, const std::vector<sycl::event>& events);    (2)
+
+}
+----
+_Effects_: Enqueues an event barrier.
+Any commands submitted after this barrier cannot begin execution until all
+commands associated with [code]#events# (and any commands associated with other
+dependent events) have completed.
+
+{note}For both overloads, if [code]#events# is empty and an event barrier has no
+other dependencies (e.g., specified by [code]#handler::depends_on#), it is ot
+required to wait for any commands unless the [code]#queue# is in-order.{endnote}
```

Comment:
I found this a little confusing because a barrier doesn't really wait for anything.  It prevents subsequent commands from starting until previous ones have finished.  Maybe it would be less confusing to say:

> If `events` is empty, and an event barrier has no other dependencies (e.g. specified by `handler::depends_on`), then this operation is a no-op.

Reply:
I agree stating it's a no-op is clearer, so I've made that change in 664a9129b3130c4cc24ea7d833bed5957eeea615.

But I think we should discuss the mental model for these functions.  The decision we make here could impact both naming and implementation.  I see two different ways to describe the semantics of these functions:

1. The `barrier` command is enqueued to the device, and the device begins executing the `barrier` command as soon as its dependencies are satisfied.  The execution of the `barrier` command is an operation which waits for all previous commands/events and then completes.  This is consistent with how other SYCL and C++ "barrier" operations work (i.e., something always waits for something else at a barrier).  A user would probably expect a command with these semantics to be in `info::event_command_status::running` status between its dependencies being satisfied and the conditions for barrier completion being satisfied.

2. The `barrier` command is enqueued to the device, but the device doesn't execute anything. The submission of the `barrier` just tells the runtime that it cannot start executing later commands until certain conditions have been satisfied.  This is consistent with how other SYCL and C++ "fence" operations work (i.e., they define ordering constraints that implementations must obey).  A user would probably expect a command with these semantics to be in `info::event_command_status::running` status for 0 time, reflecting that the device doesn't actually do anything.

I don't know which of these is closest to the behavior of real implementations.  But if you think it's closer to 2), then we should consider renaming these functions `command_fence` and `event_fence`.

Reply:
Regarding the implementation, I think we should allow either (1) or (2).  I see (2) as an optimization of (1).  In  general, I think it should be permissible for the implementation to optimize away the "running" phase of any command if it doesn't need to be done, so there is never any guarantee that an event's status will be `running` at some point.  In any case, this isn't observable because there's no way to guarantee that the host calls `event::get_info` at just the right time to observe the `running` status.

Regarding the name, I'd defer to your judgement.  It seems to me, though, that the spec currently uses the word "barrier" to mean some sort of synchronization between work-item execution while it uses the word "fence" to mean observability of memory operations.  It seems like this command is closer in meaning to "barrier" than to "fence".

Reply:
Good point about the observability.

I'm not sure what my preference is regarding naming, and I could go either way.  Part of me even wonders if it should be `command_fence` and `event_barrier`!  I'll give it some more thought.

Reply:
My original concern is addressed, so I'm marking this thread as "resolved".  If you want to let others comment on the naming, feel free to un-resolve it, or start a new comment just on the naming.